### ISSUE TITLE
Passing Additional Arguments to Model 

### ISSUE THREAD

**Issue by pindapuj** [bethgelab/foolbox#582]
Hi,

Thanks for the great package! My model takes two arguments (a,b) in its forward function. However, I only need the gradients with respect to (a). 

I was wondering if the correct and general way of extending the interface. I started with trying to extend the PGD attack. My understanding is that first, I would override the `get_loss_function`, `value_and_grad` function, as well as the `run` function to include additional arguments.

```
def get_loss_fn(self,model,labels):
    def loss_fn(a,b):
        #if we are taking the gradient wrt to the inputs
        #then, we can adjust the edge weights
        logits = model(a=a,b=b)
        return ep.crossentropy(logits,labels).sum()

def value_and_grad(
        self, loss_fn, a,b,
    ):
        value = loss_fn(a=a,b=b)

        gradient = ep.zeros_like(a)
        for k in range(samples // 2):
            noise = ep.normal(a, shape=a.shape)

            pos_theta = a + sigma * noise
            neg_theta = a - sigma * noise

            if clip:
                pos_theta = pos_theta.clip(*bounds)
                neg_theta = neg_theta.clip(*bounds)

            pos_loss = loss_fn(a=pos_theta,b=b)
            neg_loss = loss_fn(a=neg_theta,b=b,)

            gradient += (pos_loss - neg_loss) * noise

        gradient /= 2 * sigma * 2 * samples

        return value, gradient

def run(
        self,
        model,
        a,
        b,
        criterion,
        epsilon: float,
    ):
            
        x0, restore_type = ep.astensor_(a)
        criterion_ = get_criterion(criterion)
        del a, criterion, kwargs

        if not isinstance(criterion_, Misclassification):
            raise ValueError("unsupported criterion")

        labels = criterion_.labels
        loss_fn = self.get_loss_fn(model, labels)

        if self.abs_stepsize is None:
            stepsize = self.rel_stepsize * epsilon
        else:
            stepsize = self.abs_stepsize

        x = x0

        if self.random_start:
            x = self.get_random_start(x0, epsilon)
            x = ep.clip(x, *model.bounds)
        else:
            x = x0

        for _ in range(self.steps):
            _, gradients = self.value_and_grad(loss_fn=loss_fn, a=x,b=b)
            gradients = self.normalize(gradients, x=x, bounds=model.bounds)
            x = x + stepsize * gradients
            x = self.project(x, x0, epsilon)
            x = ep.clip(x, *model.bounds)

        return restore_type(x)
```

Then, I can override with the following: 

```
attack.get_loss_fn =get_loss_fn  
attack.value_and_grad = value_and_grad
attack.run=run
```

However, it appears that I need to modify the `__call__` function. If I understand correctly, this is marked as @final in [link](https://github.com/bethgelab/foolbox/blob/ef30f7feb6746480c4a12f88754b60305a02e366/foolbox/attacks/base.py#L241) so, calling `attack(model=model,a=a,b=b, criterion=criterion,epsilon=0.01)` fails b/c `__call__` is missing 1 required positional requirement. 

Is there a more correct of extending the interface that gets around this issues without having to essential duplicate different attacks? Thoughts appreciated!

**Comment by zimmerrol** [comment]
Yes, that is correct. I think for such a use-case you'd have to create your own fork and modify the `__call__` functions as well (as you described above).
